File: Tutorials/ObjC/Download2GoHelloWorld/Download2GoHelloWorld/HelloWorld.m
//
//  ViewController.m
//  Download2GoHelloWorld-Objc
//
// This source file contains a very basic example showing how to use the Virtuoso Download Engine.
// Pay close attention to code comments marked IMPORTANT
//

@import VirtuosoUIKit;
@import VirtuosoClientDownloadEngine;

#import "HelloWorld.h"
#import "DemoPlayerViewController.h"

// ---------------------------------------------------------------------------------------------------------
// IMPORTANT:
// The following three values must be initialized, please contact support@penthera.com to obtain these keys
// ---------------------------------------------------------------------------------------------------------
static NSString* publicKey = @"replace_with_your_public_key";        // <-- change this
static NSString* privateKey = @"replace_with_your_private_key";      // <-- change this

@interface HelloWorld () <UIDownloadButtonDelegate>

//
// MARK: Instance data
//
@property (nonatomic, strong) NSString *exampleAssetID;
@property (nonatomic, strong) VirtuosoAsset *exampleAsset;

//
// MARK: Outlets
//
@property (weak, nonatomic) IBOutlet UIAssetStatusLabel  *statusLabel;
@property (weak, nonatomic) IBOutlet UIDownloadButton    *downloadBtn;
@property (weak, nonatomic) IBOutlet UIPlayButton        *playBtn;
@property (weak, nonatomic) IBOutlet UIAssetProgressView *statusProgressBar;

@end

@implementation HelloWorld

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    // Disable the buttons until we've started up and have restored
    // any download state from previous launches.
    self.downloadBtn.enabled = false;
    self.playBtn.enabled = false;
    
    // Set the download button delegate
    self.downloadBtn.delegate = self;
    
    self.statusLabel.text = @"Starting Engine...";
    
    // Important:
    // AssetID should be unique for each asset across your video catalog
    self.exampleAssetID = @"tears-of-steel-asset-1";
    
    // Backplane permissions require a unique user ID to properly function.
    // Production code that needs user-device tracking or download permissions
    // will need a unique customer ID. For demonstation purposes only, we use
    // the device name here.
    NSString* userName = UIDevice.currentDevice.name;
    
    //
    // Create the engine confuration
    VirtuosoEngineConfig* config = [[VirtuosoEngineConfig alloc]initWithUser:userName
                                                                   publicKey:publicKey
                                                                  privateKey:privateKey];
    
    if (!config)
    {
        UIAlertController* alert = [UIAlertController alertControllerWithTitle:@"Setup Required"
                                                                       message:@"Please contact support@penthera.com to setup the backplaneUrl, publicKey, and privateKey"
                                                                preferredStyle:UIAlertControllerStyleAlert];
        UIAlertAction* ok = [UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
            exit(0);
        }];
        [alert addAction:ok];
        [self presentViewController:alert animated:TRUE completion:nil];
        return;
    }
    
    //
    // Start the Engine
    // This method will execute async, the callback will happen on the main-thread.
    [VirtuosoDownloadEngine.instance startup:config startupCallback:^(kVDE_EngineStartupCode status) {

        if (status == kVDE_EngineStartupSuccess) {

            NSLog(@"Startup succeeded.");
            self.statusLabel.text = @"Ready to download";

            // Load existing asset if we already created one in a previous run. Do this in the background to
            // prevent blocking UI
            dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_BACKGROUND, 0), ^{
                self.exampleAsset = [VirtuosoAsset assetWithAssetID:self.exampleAssetID availabilityFilter:NO];
            });
            
            // The download button, as configured, handles both starting a download
            // and deleting the existing download.  Whatever the state of the process,
            // once the engine has started, it's safe to enable it.
            self.downloadBtn.enabled = true;

        } else {
            NSLog(@"Startup encountered error.");
        }
    }];
}

// MARK: Click Handlers

- (IBAction)playBtnClicked:(UIButton *)sender
{
    // Microsoft Smooth Streaming (HSS) requires a proprietary player and
    // is not within scope of this demo.  Ignore it here.
    if( _exampleAsset.type != kVDE_AssetTypeHSS )
    {
        UIViewController* player = [DemoPlayerViewController new];
        
        // Play the asset using the downloaded (local) data.
        [_exampleAsset playUsingPlaybackType:kVDE_AssetPlaybackTypeLocal
                           andPlayer:(id<VirtuosoPlayer>)player
                           onSuccess:^ {
                               // Present the player
                               [self presentViewController:player animated:YES completion:nil];
                           }
                              onFail:^ {
                                  self.exampleAsset = nil;
                                  NSLog(@"Can't play video!");
                              }];
    }
}

- (VirtuosoAsset*)startAssetDownload
{
    // NOTE: This delegate handler will be called on a background thread.
    
    // If we don't already have an asset, then create one and download it. Otherwise, return the asset we already have.
    if( self.exampleAsset == nil )
    {
        // Important:
        // Create asset configuration object
        VirtuosoAssetConfig* assetConfig = [[VirtuosoAssetConfig alloc]initWithURL:@"https://virtuoso-demo-content.s3.amazonaws.com/tears/index.m3u8"
                                                                           assetID:self.exampleAssetID
                                                                       description:@"Tears of Steel"
                                                                              type:kVDE_AssetTypeHLS];
        if (!assetConfig)
        {
            NSLog(@"create config failed");
            return nil;
        }
        
        // Automatically adds to queue.
        VirtuosoAsset* asset = [VirtuosoAsset assetWithConfig:assetConfig];
        self.exampleAsset = asset;
    }

    return self.exampleAsset;
}

- (void)assetDeleted
{
    self.statusLabel.text = @"Deleted";
    
    // Clear the asset (thus clearing it from all UI elements) so we're ready to download again.
    self.exampleAsset = nil;
}

// Convenience method to hook up and reset the asset with the UI elements properly.
- (void)setExampleAsset:(VirtuosoAsset *)exampleAsset
{
    _exampleAsset = exampleAsset;
    self.downloadBtn.asset = self.exampleAsset;
    self.statusProgressBar.asset = self.exampleAsset;
    self.playBtn.asset = self.exampleAsset;
    self.statusLabel.asset = self.exampleAsset;
}

@end

File: Tutorials/Swift/Download2GoHelloWorld/Download2GoHelloWorld/HelloWorld.swift
//
//  ViewController.swift
//  Download2GoHelloWorld
//
//  Created by Penthera on 11/15/18.
//  Copyright © 2018 penthera. All rights reserved.
//

import UIKit

class HelloWorld: UIViewController, VirtuosoDownloadEngineNotificationsDelegate
{
    // ---------------------------------------------------------------------------------------------------------
    // IMPORTANT:
    // The following three values must be initialzied, please contact support@penthera.com to obtain these keys
    // ---------------------------------------------------------------------------------------------------------
    let backplaneUrl = "replace_with_your_backplane_url"
    let publicKey = "replace_with_your_public_key"
    let privateKey = "replace_with_your_private_key"
    //
    // MARK: Instance data
    //
    var exampleAsset: VirtuosoAsset?
    var downloadEngineNotifications: VirtuosoDownloadEngineNotificationManager!
    
    @IBOutlet weak var statusLabel: UILabel!
    @IBOutlet weak var playBtn: UIButton!
    @IBOutlet weak var downloadBtn: UIButton!
    @IBOutlet weak var deleteBtn: UIButton!
    @IBOutlet weak var statusProgressBar: UIProgressView!
    
    override func viewDidLoad()
    {
        super.viewDidLoad()
        
        // download engine update listener
        downloadEngineNotifications = VirtuosoDownloadEngineNotificationManager.init(delegate: self)
        self.statusLabel.text = "Starting Engine..."
        
        //
        // Enable the Engine
        VirtuosoDownloadEngine.instance().enabled = true;
        
        // Backplane permissions require a unique user-id for the full range of captabilities support to work
        // Production code that needs this will need a unique customer ID.
        // For demonstation purposes only, we use the device name
        let userName = UIDevice.current.name

        //
        // Create the engine confuration
        guard let config = VirtuosoEngineConfig(user: userName,
                                                backplaneUrl: self.backplaneUrl,
                                                publicKey: self.publicKey,
                                                privateKey: self.privateKey)
        else
        {
            DispatchQueue.main.async {
                let alert = UIAlertController(title: "Setup Required", message: "Please contact support@penthera.com to setup the backplaneUrl, pulicKey, and privateKey", preferredStyle: .alert)
                alert.addAction(UIAlertAction(title: "OK", style: .default, handler: { (action) in
                    exit(0)
                }))
                self.present(alert, animated: true, completion: nil)
            }
            return
        }

        //
        // Start the Engine
        // This method will execute async, the callback will happen on the main-thread.
        VirtuosoDownloadEngine.instance().startup(config) { (status) in
            if status == .vde_EngineStartupSuccess {
                print("start succeeded")
            } else {
                print("start failed")
            }
        }
    }
    
    override func viewWillAppear(_ animated: Bool)
    {
        super.viewWillAppear(animated)
    }
    
    func displayAsset(asset: VirtuosoAsset)
    {
        let fractionComplete = asset.fractionComplete
        let status = asset.status
        var errorAdd = "";
        if( asset.downloadRetryCount > 0 ) {
            errorAdd = String(format: " (Errors: %i)", asset.downloadRetryCount)
        }
        if status != .vde_DownloadComplete && status != .vde_DownloadProcessing {
            let optionalStatus = (asset.status == .vde_DownloadInitializing) ? NSLocalizedString("Initializing: ", comment: "") : ""
            statusLabel.text = String(format: "%@%0.02f%% (%qi MB)%@", optionalStatus, fractionComplete*100.0, asset.currentSize/1024/1024, errorAdd)
            statusProgressBar.progress = Float(fractionComplete)
            downloadBtn.isEnabled = false
        }
        else if status == .vde_DownloadComplete {
            statusLabel.text = String(format: "%0.02f%% (%qi MB)", fractionComplete*100.0, asset.currentSize/1024/1024)
            statusProgressBar.progress = Float(fractionComplete)
        }
    }
    
    private func loadEngineData()
    {
        DispatchQueue.global().async
            {
            if VirtuosoDownloadEngine.instance().started {
                let downloadComplete = VirtuosoAsset.completedAssets(withAvailabilityFilter: false)
                let downloadPending = VirtuosoAsset.pendingAssets(withAvailabilityFilter: false)
                DispatchQueue.main.async {
                    if downloadComplete.count > 0 {
                        self.exampleAsset = downloadComplete.first as? VirtuosoAsset
                    }
                    else if downloadPending.count > 0 {
                        self.exampleAsset = downloadPending.first as? VirtuosoAsset
                    }
                    else {
                        self.exampleAsset = nil
                    }
                    self.refreshView()
                }
            } else {
                DispatchQueue.main.asyncAfter(deadline: .now() + 5, execute: {
                    self.loadEngineData() // try again
                })
            }
        }
    }
    
    private func refreshView()
    {
        guard let asset = self.exampleAsset else {
            self.statusLabel.text = "Ready to download"
            self.downloadBtn.backgroundColor = UIColor.black
            self.playBtn.backgroundColor = UIColor.lightGray
            self.deleteBtn.backgroundColor = UIColor.lightGray
            self.downloadBtn.isEnabled = true
            self.playBtn.isEnabled = false
            self.deleteBtn.isEnabled = false
            self.statusProgressBar.progress = 0
            return;
        }
        
        if asset.fractionComplete < 1.0 {
            statusLabel.text = "Downloading..."
        } else {
            statusLabel.text = "Ready to play"
        }
        
        VirtuosoAsset.isPlayable(asset) { (playable) in
            if playable {
                self.playBtn.isEnabled = true
                self.playBtn.backgroundColor = UIColor.black
            } else {
                self.playBtn.isEnabled = false
                self.playBtn.backgroundColor = UIColor.lightGray
            }
        }
        
        deleteBtn.isEnabled = true
        deleteBtn.backgroundColor = UIColor.black
        
        downloadBtn.isEnabled = false
        downloadBtn.backgroundColor = UIColor.lightGray
        
        statusProgressBar.progress = Float(asset.fractionComplete)
    }
    
    @IBAction func deleteBtnClicked(_ sender: UIButton)
    {
        sender.isEnabled = false   // Block reentry
        
        guard let asset = self.exampleAsset else { return }
        self.statusLabel.text = "Deleting Asset: \(asset)"

        self.exampleAsset = nil;
        asset.delete()
        loadEngineData()
    }
    
    @IBAction func playBtnClicked(_ sender: UIButton)
    {
        guard let asset = self.exampleAsset else {
            print("asset not playable yet")
            return
        }
        
        VirtuosoAsset.isPlayable(asset) { (playable) in
            if !playable { return } // not playable yet

            switch(asset.type) {
            case .vde_AssetTypeHLS, .vde_AssetTypeDASH, .vde_AssetTypeNonSegmented:
                let demoPlayer = DemoPlayerViewController()
                
                asset.play(using: .vde_AssetPlaybackTypeLocal, andPlayer: demoPlayer as VirtuosoPlayer, onSuccess: {
                    self.present(demoPlayer, animated: true, completion: nil)
                    
                }) {
                    self.exampleAsset = nil
                    print("Video is unplayble.")
                }
                break
                
            case .vde_AssetTypeHSS:
                print("Playback of HSS supported in another Tutorial")
                break
                
            @unknown default:
                break;
            }
        }
    }
    
    @IBAction func downloadBtnClicked(_ sender: UIButton)
    {
        sender.isEnabled = false // block reentrancy while we create the asset
        
        if .vde_ReachableViaWiFi != VirtuosoDownloadEngine.instance().currentNetworkStatus() {
            let alertView = UIAlertController(title: "Network Unreachable", message: "Demo requires WiFi", preferredStyle: .alert)
            alertView.addAction(UIAlertAction(title: "Ok", style: .cancel, handler: { (action) in
                sender.isEnabled = true
            }))
            self.present(alertView, animated: true, completion: nil)
            return
        }
        
        // Create the Asset on a background thread
        DispatchQueue.global(qos: .background).async {
            // Important: AssetID should be unique across your video catalog
            let yourAssetID = "tears-of-steel-asset-1"
            
            // Create asset configuration object
            guard let config = VirtuosoAssetConfig(url: "http://virtuoso-demo-content.s3.amazonaws.com/tears/index.m3u8",
                                                   assetID: yourAssetID,
                                                   description: "Tears of Steel",
                                                   type: kVDE_AssetType.vde_AssetTypeHLS) else {
                                                    print("create config failed")
                                                    sender.isEnabled = true
                                                    return
            }
            
            // Create asset and commence downloading.
            let _ = VirtuosoAsset.init(config: config)
            DispatchQueue.main.async {
                sender.isEnabled = true
                self.refreshView()
            }
        }

    }
    
    //
    // MARK: VirtuosoDownloadEngineNotificationsDelegate - required methods ONLY
    //
    
    // ------------------------------------------------------------------------------------------------------------
    //  Called whenever the Engine starts downloading a VirtuosoAsset object.
    // ------------------------------------------------------------------------------------------------------------
    func downloadEngineDidStartDownloadingAsset(_ asset: VirtuosoAsset)
    {
        displayAsset(asset: asset)
    }
    
    // ------------------------------------------------------------------------------------------------------------
    // Called whenever the Engine reports progress for a VirtuosoAsset object
    // ------------------------------------------------------------------------------------------------------------
    func downloadEngineProgressUpdated(for asset: VirtuosoAsset)
    {
        DispatchQueue.global(qos: .background).async {
            VirtuosoAsset.isPlayable(asset) { (playable) in
                if playable {
                    self.exampleAsset = asset;
                    self.refreshView()
                }
                self.displayAsset(asset: asset)
            }
        }
    }
    
    // ------------------------------------------------------------------------------------------------------------
    // Called when an asset is being processed after background transfer
    // ------------------------------------------------------------------------------------------------------------
    func downloadEngineProgressUpdatedProcessing(for asset: VirtuosoAsset)
    {
        displayAsset(asset: asset)
    }
    
    // ------------------------------------------------------------------------------------------------------------
    // Called whenever the Engine reports a VirtuosoAsset as complete
    // ------------------------------------------------------------------------------------------------------------
    func downloadEngineDidFinishDownloadingAsset(_ asset: VirtuosoAsset)
    {
        // Download is complete
        displayAsset(asset: asset)
        refreshView()
        loadEngineData()
    }
    
    // ------------------------------------------------------------------------------------------------------------
    // Called whenever an asset is added to the Engine
    // ------------------------------------------------------------------------------------------------------------
    func downloadEngineInternalQueueUpdate() {
        loadEngineData()
    }
    
    // ------------------------------------------------------------------------------------------------------------
    // Called whenever Engine start completes
    // ------------------------------------------------------------------------------------------------------------
    
    func downloadEngineStartupComplete(_ succeeded: Bool) {
        loadEngineData()
    }
}

